Put artifacts in well-known locations

ViktorHofer opened the pull request:
Fixes https://github.com/dotnet/source-build/issues/4104
Fixes https://github.com/dotnet/source-build/issues/3696
Contributes to https://github.com/dotnet/source-build/issues/4101
Contributes to https://github.com/dotnet/arcade-services/issues/3280

1. Split shipping and non-shipping packages and put them into the Arcade default locations:
-> artifacts/packages/<config>/Shipping
-> artifacts/packages/<config>/NonShipping

Put artifacts under -> `artifacts/assets/<config>`

Add the local feed for the addition package directory to NuGet.config.

2. Remove the non-shipping packages list generation now that artifacts are grouped.

3. Generate the intermediate PackageVersions* files in -> artifacts/obj/PackageVersions

4. Extract the repo symbol archives intermediates to -> artifacts/obj/Symbols/

5. Improve artifacts logging and ensuring that artifacts are created

6. Sequence targets better and apply correct DependsOnTargets

7. Small clean-up

8. Allow the poison infrastructure to analyze SBRP packages based on item metadata instead of an expected directory structure in the archive. Make the poison infrastructure analyze nuget packages directly.

ViktorHofer commented on src/SourceBuild/content/repo-projects/installer.proj:
```diff
@@ -29,7 +29,7 @@
 
     <!-- Consume the source-built Core-Setup and toolset. This line must be removed to source-build CLI without source-building Core-Setup first. -->
-    <BuildArgs>$(BuildArgs) /p:PublicBaseURL=file:%2F%2F$(SourceBuiltAssetsDir)</BuildArgs>
+    <BuildArgs>$(BuildArgs) /p:PublicBaseURL=file:%2F%2F$(ArtifactsShippingPackagesDir)</BuildArgs>
```
This line worries me a bit. I'm not sure if installer needs the transport feed as well. If it does, how would we pass that in? cc @mmitche

ViktorHofer replied:
OK the build failure is telling that we really need both local feeds.

mmitche replied:
It will need the transport feed dir too. We don't typically separate blob assets by shipping and non-shipping. Those go to a single storage locaiton. Packages go to two locations.

The reason being is mainly that we can install SDK +, add shipping feed Y, and test.

ViktorHofer commented on src/SourceBuild/content/repo-projects/Directory.Build.targets:
```diff
@@ -276,15 +289,27 @@
              StopOnFirstFailure="true" />
   </Target>
 
+  <Target Name="SetSourceBuiltSdkOverrides"
+          Condition="'@(SourceBuiltSdkOverride)' != ''">
+    <ItemGroup>
+      <!-- Set the environment variables for MSBuild to look for our additional SDK Resolvers and or our resolver to find our source-built SDKs. -->
+      <EnvironmentVariables Include="MSBUILDADDITIONALSDKRESOLVERSFOLDER=$(VSMSBuildSdkResolversDir)" />
+      <EnvironmentVariables Include="SOURCE_BUILT_SDK_ID_%(SourceBuiltSdkOverride.Group)=%(SourceBuiltSdkOverride.Identity)" />
+      <EnvironmentVariables Include="SOURCE_BUILT_SDK_VERSION_%(SourceBuiltSdkOverride.Group)=%(SourceBuiltSdkOverride.Version)" />
+      <EnvironmentVariables Condition="'%(SourceBuiltSdkOverride.Location)' != ''" Include="SOURCE_BUILT_SDK_DIR_%(SourceBuiltSdkOverride.Group)=%(SourceBuiltSdkOverride.Location)/" />
+      <EnvironmentVariables Condition="'%(SourceBuiltSdkOverride.Location)' == ''" Include="SOURCE_BUILT_SDK_DIR_%(SourceBuiltSdkOverride.Group)=$(SourceBuiltSdksDir)%(SourceBuiltSdkOverride.Identity)/" />
+    </ItemGroup>
+  </Target>
```
This target just got moved up to the build stage.

ViktorHofer commented on src/SourceBuild/content/repo-projects/Directory.Build.targets:
```diff
@@ -336,21 +362,11 @@
     <Message Importance="High" Text="See '$(RepoConsoleLogFile)' for more information." Condition="Exists('$(RepoConsoleLogFile)') and '$(MinimalConsoleLogOutput)' == 'true'" />
   </Target>
 
-  <Target Name="Build" DependsOnTargets="BuildRepoReferences;RepoBuild" />
-
-  <Target Name="Package"
-          AfterTargets="Build"
-          DependsOnTargets="
-    CopyRepoArtifacts;
-    CopyInnerBuildRestoredPackages;
-    EnsurePackagesCreated;
-    ExtractToolPackage;
-    CleanupRepo" />
-
```
These targets got combined and moved down.

ViktorHofer commented:
/azp run

MichaelSimons commented on .devcontainer/vmr-source-build/README.md:
```diff
@@ -13,7 +13,7 @@ This Codespace can help you debug the source build of .NET. In case you have run
 the VMR which is what the VMR leg in the installer PR build doing. This build takes about 45
 minutes and, after completion, produces an archived .NET SDK located in
-`/workspaces/dotnet/artifacts/x64/Release`.
+`/workspaces/dotnet/artifacts/packages/Release/Shipping`.
```
The VMR's readme needs a similar change in step 2 and 3 of [Building](https://github.com/dotnet/dotnet?tab=readme-ov-file#building)

MichaelSimons replied:
I missed that you already updated it.

MichaelSimons commented on src/SourceBuild/content/Directory.Build.props:
```diff
@@ -168,14 +171,7 @@
-    <SourceBuiltBlobFeedDir>$([MSBuild]::NormalizeDirectory('$(SharedIntermediateOutputPath)', 'blob-feed'))</SourceBuiltBlobFeedDir>
-    <SourceBuiltPackagesPath>$([MSBuild]::NormalizeDirectory('$(SourceBuiltBlobFeedDir)', 'packages'))</SourceBuiltPackagesPath>
-    <SourceBuiltAssetsDir>$([MSBuild]::NormalizeDirectory('$(SourceBuiltBlobFeedDir)', 'assets'))</SourceBuiltAssetsDir>
+    <IntermediateSymbolsRootDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsObjDir)', 'Symbols'))</IntermediateSymbolsRootDir>
```
How does the Microsoft build gather/publish symbols?  Does it want the individual symbol packages?  I ask wondering if the symbols should be places next to the packages they are for?

ViktorHofer replied:
> How does the Microsoft build gather/publish symbols? Does it want the individual symbol packages? 

Correct. Symbol packages are placed next to the non symbol packages in MSFT builds. We want symbol packages here as well but they are currently filtered out: https://github.com/dotnet/source-build/issues/4110

MichaelSimons commented on src/SourceBuild/content/Directory.Build.props:
```diff
@@ -92,6 +92,9 @@
     <ArtifactsObjDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsDir)', 'obj'))</ArtifactsObjDir>
     <ArtifactsBinDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsDir)', 'bin'))</ArtifactsBinDir>
     <ArtifactsLogDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsDir)', 'log', '$(Configuration)'))</ArtifactsLogDir>
+    <ArtifactsPackagesDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsDir)', 'packages', '$(Configuration)'))</ArtifactsPackagesDir>
```
I find it a little confusing that this is called packages when is contains more than just nupkgs.  I know this is the arcade way but...

ViktorHofer replied:
Agreed, but I'd rather stay consistent with all other arcadified repos.

mmitche replied:
Technically all other repos don't use that same layout. I wuld stick with packages == nuget packages and a sibling 'assets' == everything else.

ViktorHofer replied:
Chatted with Matt offline. I will move the assets out.

ViktorHofer commented on src/SourceBuild/content/repo-projects/installer.proj:
```diff
@@ -29,7 +29,7 @@
     <BuildArgs Condition="'$(TargetOS)' != 'windows'">$(BuildArgs) /p:CoreSetupRid=$(TargetRid)</BuildArgs>
 
     <!-- Consume the source-built Core-Setup and toolset. This line must be removed to source-build CLI without source-building Core-Setup first. -->
```
No idea... I will leave it as is for now.

NikolaMilosavljevic commented on src/SourceBuild/content/eng/build.sourcebuild.targets:
```diff
@@ -15,45 +15,17 @@
     <MSBuild Projects="$(RepoProjectsDir)$(RootRepo).proj" Targets="WritePrebuiltUsageData;ReportPrebuiltUsage" />
   </Target>
 
-  <Target Name="DiscoverSymbolsTarballs"
```
Hmm, we need to extract original `symbols.<repo>` packages. Is this now done in a different target/file?

ViktorHofer replied:
Yes. That's now in repo-projects/Directory.Build.targets.

MichaelSimons commented on eng/pipelines/templates/jobs/vmr-build.yml:
```diff
@@ -348,7 +347,9 @@ jobs:
         CopyWithRelativeFolders "artifacts/"                 $targetFolder "*.binlog"
         CopyWithRelativeFolders "artifacts/"                 $targetFolder "*.log"
-        CopyWithRelativeFolders "artifacts/prebuilt-report/" $targetFolder
+        if ("${{ parameters.buildSourceOnly }}" -eq $True) {
```
Is this even necessary at all on Windows?  Would we ever support a source-only windows build?

ViktorHofer commented on eng/pipelines/vmr-build.yml:
```diff
@@ -56,9 +56,9 @@ stages:
       ${{ if eq(variables.isSourceOnlyBuild, 'true') }}:
-        scope: ultralite
+        scope: full
       ${{ else }}:
-        scope: lite
+        scope: full
```
Temp change to queue a full build

ViktorHofer commented:
The failures are due to https://github.com/dotnet/source-build/issues/4111

MichaelSimons commented on src/SourceBuild/content/Directory.Build.props:
```diff
@@ -166,14 +169,8 @@
-    <SourceBuiltPackagesPath>$([MSBuild]::NormalizeDirectory('$(SourceBuiltBlobFeedDir)', 'packages'))</SourceBuiltPackagesPath>
-    <SourceBuiltAssetsDir>$([MSBuild]::NormalizeDirectory('$(SourceBuiltBlobFeedDir)', 'assets'))</SourceBuiltAssetsDir>
+    <IntermediateSymbolsRootDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsObjDir)', 'Symbols'))</IntermediateSymbolsRootDir>
+    <ArtifactsAssetsDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsDir)', 'assets', '$(Configuration)'))</ArtifactsAssetsDir>
```
Consider grouping with the other Artifacts*Dir properties.

ViktorHofer replied:
We don't want to put anything in [here](https://github.com/dotnet/dotnet/blob/f1bfb8c1cfee38cc59bb336e1fb9b37dd47ebb16/Directory.Build.props#L71-L119) that isn't defined by the Arcade SDK.

MichaelSimons replied:
Fair enough, I missed that context in the clipped diff view.

ViktorHofer commented on eng/pipelines/templates/stages/vmr-build.yml:
```diff
@@ -122,7 +122,7 @@ stages:
             buildFromArchive: false            # 🚫
             buildSourceOnly: true              # ✅
-            enablePoison: false                # 🚫
+            enablePoison: true                 # 🚫
```
Temporary change

ViktorHofer commented:
@MichaelSimons @NikolaMilosavljevic I fixed the leak detection issue and updated the branch with latest main. The remaining failure exists in main as well. This PR is now ready. PTAL.

MichaelSimons commented on src/SourceBuild/content/eng/build.sourcebuild.targets:
```diff
@@ -111,21 +83,22 @@
           Outputs="$(BaseIntermediateOutputPath)ReportPoisonUsage.complete" >
     <ItemGroup>
-      <FinalCliTarball Include="$(SharedOutputPath)**/*$(ArchiveExtension)" />
+      <!-- Exclude the Private.SourceBuilt.Artifacts archive from poison usage scan. -->
```
Would it make sense to move this comment to be directly above the line that is doing the removal?
